### PR TITLE
Make assert! ensure the macro is parsed completely

### LINKED ISSUE #60024: assert! allows invalid trailing tokens in expression
```rust
fn main() {
    assert!(true some extra junk, "whatever");  
    assert!(true some extra junk);                    
}
```

([Playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=ff19f9ba70a1c5b936b658a4914182a7))

**Output:**
```
   Compiling playground v0.0.1 (/playground)
    Finished dev [unoptimized + debuginfo] target(s) in 0.54s
     Running `target/debug/playground`
```

**Version:**
* 1.34.0
* 1.35.0 2019-04-15 nightly (2975a3c4befa8ad610da)

**Expected:**
```
error: expected one of `.`, `;`, `?`, or an operator, found `some`
 --> src/main.rs:2:18
  |
2 |     assert!(true some extra junk, "whatever");
  |                  ^^^^ expected one of `.`, `;`, `?`, or an operator here
```

**Original report:**
https://www.reddit.com/r/rust/comments/bdbtpr/hey_rustaceans_got_an_easy_question_ask_here/el1gm0h/

### PR DESCRIPTION
**Pull request by rasendubi**
Fixes https://github.com/rust-lang/rust/issues/60024

### CODE REVIEW

**Comment by jonas-schievink** [comment]
Is there any reason why `assert!` is built into the compiler instead of just being defined normally in libcore?

**Comment by petrochenkov** [comment]
@jonas-schievink 
It was moved from libcore to compiler as a first step in implementing https://github.com/rust-lang/rust/issues/44838 (the second step never happened though).

**Comment by alexcrichton** [comment]
Thanks @rasendubi! To be double extra sure I think we'll want to run this through crater to ensure we don't break too many crates by accident, so...

@bors: try

File: src/test/ui/macros/assert-trailing-junk.rs [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+fn main() {
+    assert!(true some extra junk, "whatever");
+    //~^ ERROR expected one of
+
+    assert!(true some extra junk);
+    //~^ ERROR expected one of
+
+    assert!(true, "whatever" blah);
+    //~^ ERROR no rules expected
```
Comment (estebank) [inline src/test/ui/macros/assert-trailing-junk.rs]: Is `assert!(foo, bar, baz)` supposed to be accepted? I think you'll need to also check how many expressions have been found so far and error if more than 2.

**Comment by alexcrichton** [comment]
@craterbot run start=master#efe2f32a6b8217425f361ec7c206910c611c03ee end=try#78b0bd8441ef961f6475e3d47fe3bca38cb83432 mode=check-only

File: src/test/ui/macros/assert-trailing-junk.rs (hunk 1 above)
Comment (rasendubi) [reply src/test/ui/macros/assert-trailing-junk.rs]: I thought about this when implementing, but it is accepted. Second to last arguments represent format string + its arguments.

```rust
assert!(foo == 3, "foo is expected to be 3 but is {}", foo);
```
produces
```
thread 'main' panicked at 'foo is expected to be 3 but is 2', test.rs:3:5
```

Comment (estebank) [reply src/test/ui/macros/assert-trailing-junk.rs]: Ah! I see, you're right. Disregard me.

**Comment by rasendubi** [comment]
The crater agent is unreachable for a day now. Is that normal?

Last heartbeat: 2019-04-20 06:26:03 UTC

**Comment by estebank** [comment]
CC @pietroalbini ^^

**Comment by pietroalbini** [comment]
@craterbot retry

The Crater machine crashed, now it should work again.

**Comment by alexcrichton** [comment]
That looks like an unfortuante amount of breakage :(

Given all that I think that the best thing to do might be to switch this to a warning, and then perhaps send some PRs to upstream crates to fix the issues? After awhile we can then switch to an error

**Comment by estebank** [comment]
@alexcrichton a significant amount of these (8) seem to be coming from `mop-structs` having the equivalent of `assert!(a == b "error message")` (note missing comma), which we could explicitly detect and unconditionally warn on. `q1tsim` and `carrier` have the same kind of error.

`static-linkedlist` has `assert!(error;)`.

[bluenrg](https://crater-reports.s3.amazonaws.com/pr-60039/try%2378b0bd8441ef961f6475e3d47fe3bca38cb83432/reg/bluenrg-0.0.3/log.txt), [cargo-generate](https://crater-reports.s3.amazonaws.com/pr-60039/try%2378b0bd8441ef961f6475e3d47fe3bca38cb83432/reg/cargo-generate-0.3.0/log.txt) (and [it's transitive error](https://crater-reports.s3.amazonaws.com/pr-60039/try%2378b0bd8441ef961f6475e3d47fe3bca38cb83432/gh/ashleygwilliams.cargo-generate/log.txt)) and [b-r-oleary.passwords](https://crater-reports.s3.amazonaws.com/pr-60039/try%2378b0bd8441ef961f6475e3d47fe3bca38cb83432/gh/b-r-oleary.passwords/log.txt) were a bit more bizarre on first look without context, but looking a bit deeper I see they are similar to `static-linkedlist`:

https://github.com/danielgallagher0/bluenrg/blob/a47f1f57277995937e1cf6ae8c8f58d8e7254187/tests/l2cap.rs#L64-L69

https://github.com/ashleygwilliams/cargo-generate/blob/54a85ed98ee4ad06c8228dd60ac070ff483291dd/tests/integration/basics.rs#L447-L450

https://github.com/b-r-oleary/passwords/blob/c85100139d8301f72c6948b672e98247571c8a51/src/generators/phrase.rs#L195-L200

I think if we accept trailing `;` and trailing `&'static str`s in assert, potentially linting about these, we can merge the fix.

File: src/libsyntax_ext/assert.rs [hunk 2]
```diff
@@ -86,5 +86,12 @@ fn parse_assert<'a>(
         } else {
             None
         },
-    })
+    };
+
+    if parser.token != token::Eof {
+        parser.expect_one_of(&[], &[])?;
```
Comment (estebank) [inline src/libsyntax_ext/assert.rs]: You can add a check here for `parser.token == token::Semi` and swallow it or warn about it, depending on which way we want to go about it.

You can also try parser.parse_tokens() and modify `assert.custom_message` with the result, and complain about it then.

**Comment by rasendubi** [comment]
Updated the code to:
1. warn and skip trailing semicolon
2. warn about missing semicolon and trying to parse the rest as `custom_message`

Going to submit PRs to upstream crates tomorrow.

**Comment by estebank** [comment]
@craterbot run start=master#e305df1846a6d985315917ae0c81b74af8b4e641 end=try#99a7498759a22b5684caf798a98b7e90ad9d2443 mode=check-only

**Comment by kennytm** [comment]
@craterbot ping

**Comment by kennytm** [comment]
@craterbot run mode=check-only

**Comment by rasendubi** [comment]
PRs to upstream crates:
- https://github.com/danielgallagher0/bluenrg/pull/2
- https://github.com/ashleygwilliams/cargo-generate/pull/157
- https://gitlab.com/c410-f3r/mop/merge_requests/1
- https://github.com/Q1tBV/q1tsim/pull/1
- https://github.com/arkap/rs-static-linkedlist/pull/1
- https://github.com/aep/carrier-legacy/pull/1
- https://github.com/aep/cr2/pull/2
- https://github.com/b-r-oleary/passwords/pull/1
- https://github.com/ecsumed/wrench/pull/1
- https://github.com/focusaurus/cody/pull/1
- https://github.com/wduquette/bonaventure/pull/2

**Comment by rasendubi** [comment]
The one regression reported by crater does not seem relevant to the change. From [logs](https://crater-reports.s3.amazonaws.com/pr-60039-1/try%2399a7498759a22b5684caf798a98b7e90ad9d2443/reg/rsc-1.2.0/log.txt):
```
[INFO] [stderr] error[E0061]: this function takes 2 parameters but 0 parameters were supplied
[INFO] [stderr]   --> src/bin/main.rs:29:24
[INFO] [stderr]    |
[INFO] [stderr] 29 |     let mut computer = Computer::new();
[INFO] [stderr]    |                        ^^^^^^^^^^^^^^^ expected 2 parameters
[INFO] [stderr] 
[INFO] [stderr] error[E0061]: this function takes 2 parameters but 1 parameter was supplied
[INFO] [stderr]   --> src/bin/main.rs:43:15
[INFO] [stderr]    |
[INFO] [stderr] 43 |         match tokenize(&buffer) {
[INFO] [stderr]    |               ^^^^^^^^^^^^^^^^^ expected 2 parameters
[INFO] [stderr] 
[INFO] [stderr] error: aborting due to 2 previous errors
```

**Comment by alexcrichton** [comment]
@bors: r+

This looks like a great compromise, thanks @rasendubi!

**Comment by rasendubi** [comment]
This is my first contribution to rust (besides some minor docs improvements). Thank you guys for all the support! :tada:

**Pull request merged**